mjpc/tasks/humanoid/tracking/task.cc

```diff
@@ -0,0 +1,157 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tasks/humanoid/tracking/task.h"
+
+#include <array>
+#include <iostream>
+#include <map>
+
+#include <mujoco/mujoco.h>
+#include "utilities.h"
+
+
+namespace mjpc {
+
+// ------------- Residuals for humanoid tracking task -------------
+//   Number of residuals: TODO(hartikainen)
+//     Residual (0): TODO(hartikainen)
+//   Number of parameters: TODO(hartikainen)
+//     Parameter (0): TODO(hartikainen)
+// ----------------------------------------------------------------
+  void Humanoid::Tracking::Residual(const double* parameters, const mjModel* model,
+                                    const mjData* data, double* residual) {
+    int counter = 0;
+
+    // float fps = 30.0;
+    // int step_index = std::min((int) (data->time * fps), (model->nkey) - 1);
+
+    mju_copy(&residual[counter], data->ctrl, model->nu);
+    counter += model->nu;
+
+    // ----- joint velocity ----- //
+    mju_copy(residual + counter, data->qvel, model->nv);
+    counter += model->nv;
+
+    std::array<std::string, 16> body_names = {
+      "pelvis", "head", "ltoe", "rtoe", "lheel", "rheel", "lknee", "rknee",
+      "lhand", "rhand", "lelbow", "relbow", "lshoulder", "rshoulder", "lhip",
+      "rhip",
+    };
+
+    for (const auto& body_name : body_names) {
+      std::string mocap_body_name = "mocap[" + body_name + "]";
+      std::string pos_sensor_name = "tracking_pos[" + body_name + "]";
+      int body_mocapid = model->body_mocapid[mj_name2id(model, mjOBJ_BODY, mocap_body_name.c_str())];
+      assert(0 <= body_mocapid);
+
+      double mocap_body_pos[3];
+      mju_copy3(mocap_body_pos, data->mocap_pos + 3 * body_mocapid);
+      // mju_copy3(mocap_body_pos, model->key_mpos + model->nmocap * 3 * step_index + 3 * body_mocapid);
+
+      double* sensor_pos = mjpc::SensorByName(model, data, pos_sensor_name.c_str());
+
+      mju_sub3(&residual[counter],
+               mocap_body_pos,
+               sensor_pos);
+      counter += 3;
+    }
+
+    for (const auto& body_name : body_names) {
+      std::string mocap_body_name = "mocap[" + body_name + "]";
+      std::string linvel_sensor_name = "tracking_linvel[" + body_name + "]";
+      int body_mocapid = model->body_mocapid[
+        mj_name2id(model, mjOBJ_BODY, mocap_body_name.c_str())];
+      assert(0 <= body_mocapid);
+
+      double mocap_body_linvel[3];
+      mju_copy3(mocap_body_linvel, data->mocap_quat + 3 * body_mocapid);
+      // double current_mocap_body_pos[3];
```

Comment:
For whatever reason doing this here seemed to break everything when we tested this out with @thowell earlier. Things work fine if we compute the finite-difference velocities (exactly the same way) in `Transition` and store them in `mocap_quat` and use them here. If anyone has ideas what might be causing this, let me know!

Reply:
@nimrod-gileadi any ideas? I am guessing we were trying something silly in C++...

Reply:
Only thing I noticed is that step_index+1 might be equal to model->nkey, which would overflow the array, which isn't a problem in the Transition implementation.